### PR TITLE
(PUP-10844) Warnings for unavailable servers

### PR DESCRIPTION
**Pull request by Dorin-Pleava**
When processing server_list and trying to find a functional server,
throw warnings when it cannot connect to servers from server_list and
if there is no functional server in server_list, then throw the initial
error.

### CODE REVIEW

File: lib/puppet/http/resolver/server_list.rb [hunk 1]
```diff
@@ -63,7 +63,12 @@ def resolve(session, name, ssl_context: nil, canceled_handler: nil)
         Puppet.log_exception(detail, _("Puppet server %{host}:%{port} is unavailable: %{code} %{reason}") %
                              { host: service.url.host, port: service.url.port, code: detail.response.code, reason: detail.response.reason })
       rescue Puppet::HTTP::HTTPError => detail
-        Puppet.log_exception(detail, _("Unable to connect to server from server_list setting: %{detail}") % {detail: detail})
+        if server.equal?(@server_list_setting.value.last)
```
Comment (joshcooper) [inline lib/puppet/http/resolver/server_list.rb:66]: It's possible, though unlikely that a server might appear multiple times in list, in which case we would call `log_exception` multiple times:

```
server_list: a, b, a
```
Might be clearer to pass the index when iterating so we know when we've reached the end?

```ruby
@server_list_setting.value.each_with_index do |server, i|
  ...
  rescue Puppet::HTTP::HTTPError => detail
    if i < @server_list_setting.value.length - 1
     # try next server case
   else
    # reached end of list case
  end
end
```

**Comment by joshcooper** [comment]
nit typo in the subject, should be "Warnings for .."

**Comment by vchepkov** [comment]
Should be Puppet::HTTP::ResponseError handled similar way? agent will continue to the next server if initial server went down, right?

**Comment by Dorin-Pleava** [comment]
Yep, it seems like when the server sends a Puppet::HTTP::ResponseError, it will try with the next server from server_list.
Added the same logic to Puppet::HTTP::ResponseError.

**Review by joshcooper — APPROVED** [review]
LGTM though I think we want to target `6.x` and you'll want to rebase on upstream to get the win32-dir fix

**Comment by joshcooper** [comment]
I verified that the agent prints an error and fails the run if it exhausts the server_list due to `connect` failing, e.g. due to DNS, no route to host, etc:

```
$ bx puppet agent -t --server_list puppet.example.com                               
Info: csr_attributes file loading from /home/josh/.puppetlabs/etc/puppet/csr_attributes.yaml
Info: Creating a new SSL certificate request for localhost
Info: Certificate Request fingerprint (SHA256): 1F:9C:F3:F7:A5:89:FF:F0:AA:55:11:EA:13:48:A3:F2:AB:7E:60:CC:66:9C:BA:84:0E:C0:EE:BA:09:67:14:C0
Error: Unable to connect to server from server_list setting: Request to https://puppet.example.com:8140/status/v1/simple/master failed after 0.112 seconds: Failed to open TCP connection to puppet.example.com:8140 (getaddrinfo: Name or service not known)
Wrapped exception:
Failed to open TCP connection to puppet.example.com:8140 (getaddrinfo: Name or service not known)
Error: No more routes to ca
Error: Could not run: No more routes to ca
```

Or if the puppetserver's HTTP response is not successful:

```
$ bx puppet agent -t --server_list localhost
Error: Puppet server localhost:8140 is unavailable: 500 Error
```

If the server does eventually find a "good" server, then only warnings are printed in the `connect` case:

```
$ bx puppet agent -t --server_list puppet.example.com,sized-engraving.delivery.puppetlabs.net
Warning: Unable to connect to server from server_list setting: Request to https://puppet.example.com:8140/status/v1/simple/master failed after 0.099 seconds: Failed to open TCP connection to puppet.example.com:8140 (getaddrinfo: Name or service not known) Trying with next server from server_list.
Info: Creating a new RSA SSL key for localhost
Info: csr_attributes file loading from /home/josh/.puppetlabs/etc/puppet/csr_attributes.yaml
Info: Creating a new SSL certificate request for localhost
Info: Certificate Request fingerprint (SHA256): FF:4C:04:78:33:25:38:4E:B2:A6:48:67:E1:18:44:61:22:CA:9C:64:26:32:6E:66:78:18:B5:30:19:FF:0A:4E
Info: Certificate for localhost has not been signed yet
```

And the same in the non-2xx HTTP response case:

```
$ bx puppet agent -t --server_list localhost,sized-engraving.delivery.puppetlabs.net
Warning: Puppet server localhost:8140 is unavailable: 503 Busy Trying with next server from server_list.
Info: Creating a new RSA SSL key for localhost
Info: csr_attributes file loading from /home/josh/.puppetlabs/etc/puppet/csr_attributes.yaml
Info: Creating a new SSL certificate request for localhost
Info: Certificate Request fingerprint (SHA256): 00:F2:A0:EC:B1:79:7C:09:B0:06:59:2C:C1:0B:6E:48:13:8E:4D:08:52:D2:09:8C:AD:DA:90:04:48:33:D6:0F
Info: Certificate for localhost has not been signed yet
```

**Pull request merged**